subsys/lorawan/services/multicast.c

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright (c) 2022-2024 Martin Jäger <martin@libre.solar>
+ * Copyright (c) 2022-2024 tado GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "lorawan_services.h"
+#include "../lw_priv.h"
+
+#include <LoRaMac.h>
+#include <zephyr/lorawan/lorawan.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/random/random.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <stdio.h>
+
+LOG_MODULE_REGISTER(lorawan_multicast, CONFIG_LORAWAN_SERVICES_LOG_LEVEL);
+
+/**
+ * Version of LoRaWAN Remote Multicast Setup Specification
+ *
+ * This implementation only supports TS005-1.0.0.
+ */
+#define MULTICAST_PACKAGE_VERSION 1
+
+/* maximum length of multicast answers */
+#define MAX_MULTICAST_ANS_LEN 5
+
+#define MULTICAST_SETTINGS_BASE "lorawan/multicast"
+
+enum multicast_commands {
+	MULTICAST_CMD_PKG_VERSION = 0x00,
+	MULTICAST_CMD_MC_GROUP_STATUS = 0x01,
+	MULTICAST_CMD_MC_GROUP_SETUP = 0x02,
+	MULTICAST_CMD_MC_GROUP_DELETE = 0x03,
+	MULTICAST_CMD_MC_CLASS_C_SESSION = 0x04,
+	MULTICAST_CMD_MC_CLASS_B_SESSION = 0x05,
+};
+
+struct multicast_context {
+	struct k_work_delayable session_start_work;
+	struct k_work_delayable session_stop_work;
+};
+
+static struct multicast_context ctx[LORAMAC_MAX_MC_CTX];
+
+static void multicast_session_start(struct k_work *work)
+{
+	int ret;
+
+	ret = lorawan_services_class_c_start();
+	if (ret < 0) {
+		LOG_WRN("Failed to switch to class C: %d. Retrying in 1s.", ret);
+		lorawan_services_reschedule_work(k_work_delayable_from_work(work), K_SECONDS(1));
+	}
+}
+
+static void multicast_session_stop(struct k_work *work)
+{
+	int ret;
+
+	ret = lorawan_services_class_c_stop();
+	if (ret < 0) {
+		LOG_WRN("Failed to revert to class A: %d. Retrying in 1s.", ret);
+		lorawan_services_reschedule_work(k_work_delayable_from_work(work), K_SECONDS(1));
+	}
+}
+
+/**
+ * Schedule Class C session if valid timing is found
+ *
+ * @returns time to start (negative in case of missed start)
+ */
+static int32_t multicast_schedule_class_c_session(uint8_t id, uint32_t session_time,
+						  uint32_t session_timeout)
+{
+	uint32_t current_time;
+	int32_t time_to_start;
+	int err;
+
+	err = lorawan_clock_sync_get(&current_time);
+	time_to_start = session_time - current_time;
+
+	if (err != 0 || time_to_start > 0xFFFFFF) {
+		LOG_ERR("Clocks not synchronized, cannot schedule class C session");
+
+		/* truncate value to indicates that clocks are out of sync */
+		time_to_start = 0xFFFFFF;
+	} else if (time_to_start >= 0) {
+		LOG_DBG("Starting class C session in %d s", time_to_start);
+
+		lorawan_services_reschedule_work(&ctx[id].session_start_work,
+						 K_SECONDS(time_to_start));
+
+		lorawan_services_reschedule_work(&ctx[id].session_stop_work,
+						 K_SECONDS(time_to_start + session_timeout));
+	}
+
+	return time_to_start;
+}
+
+static void multicast_package_callback(uint8_t port, bool data_pending, int16_t rssi, int8_t snr,
+				       uint8_t len, const uint8_t *rx_buf)
+{
+	uint8_t tx_buf[3 * MAX_MULTICAST_ANS_LEN];
```

Comment:
Where does the 3 come from here?
Is it just the maximum number of requests you've observed in a single packet?

Reply:
It was just a random number to reserve a bit more space than just for one message. The standard states "A message MAY carry more than one command.", but there is no limit specified, which is a bit annoying.

I have now created a define for that and explained the decision in a comment.

We may want to turn this into a Kconfig if you prefer, even though I don't think anyone will ever need more than 3 messages.